File: src/SnoDb/SnoCollection.vb
Imports System.IO
Imports System.IO.Compression
Imports System.Collections.Concurrent
Imports Newtonsoft.Json

Public Class SnoCollection(Of T, TId)
    Implements ISnoCollection(Of T)

    Public ReadOnly Property Database As SnoDatabase Implements ISnoCollection.Database
    Public ReadOnly Property Name As String Implements ISnoCollection.Name
    Private ReadOnly Property IdSelector As Func(Of T, TId)
    Private ReadOnly Property Items As ConcurrentDictionary(Of TId, T) = New ConcurrentDictionary(Of TId, T)()
    Private ReadOnly Property SaveActions As ConcurrentDictionary(Of TId, ISaveAction) = New ConcurrentDictionary(Of TId, ISaveAction)()

    Public Sub New(database As SnoDatabase, name As String, idSelector As Func(Of T, TId))
        Me.Database = database
        Me.Name = name
        Me.IdSelector = idSelector
    End Sub

    Public Sub AddOrReplace(item As T) Implements ISnoCollection(Of T).AddOrReplace
        Dim id = Me.IdSelector(item)
        Me.Items(id) = item

        Dim itemPath = Me.GetItemPath(item)
        Me.SaveActions(id) = New AddOrReplaceSaveAction(itemPath, item)
    End Sub

    Public Sub Remove(item As T) Implements ISnoCollection(Of T).Remove
        Dim id = Me.IdSelector(item)
        Me.Items.TryRemove(id, Nothing)

        Dim itemPath = Me.GetItemPath(item)
        Me.SaveActions(id) = New RemoveSaveAction(itemPath)
    End Sub

    Public Function Query() As IQueryable(Of T) Implements ISnoCollection(Of T).Query
        Return Me.Items.Values.AsQueryable()
    End Function

    Private Function GetItemPath(item As T) As String Implements ISnoCollection(Of T).GetItemPath
        Dim id = Me.IdSelector(item)

        Return $"{Me.Name}/{id}.json"
    End Function

    Public Async Function LoadAsync(entries As IEnumerable(Of ZipArchiveEntry)) As Task Implements ISnoCollection(Of T).LoadAsync
        For Each entry In entries
            Dim itemJson As String
            Using reader = New StreamReader(entry.Open())
                itemJson = Await reader.ReadToEndAsync()
            End Using

            Dim item = JsonConvert.DeserializeObject(Of T)(itemJson)
            Dim id = Me.IdSelector(item)
            Me.Items(id) = item
        Next
    End Function

    Public Async Function SaveAsync(archive As ZipArchive) As Task Implements ISnoCollection.SaveAsync
        For Each saveAction In Me.SaveActions.Values
            Await saveAction.SaveAsync(archive)
        Next
    End Function

End Class

File: src/SnoDb/SnoCollection.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Collections.Concurrent;
using Newtonsoft.Json;
using Ionic.Zip;

namespace Symmex.SnoDb
{
    public class SnoCollection<T, TId> : ISnoCollection<T>
    {
        public SnoDatabase Database { get; }
        public string Name { get; }

        private Func<T, TId> IdSelector { get; }
        private ConcurrentDictionary<TId, T> Items = new ConcurrentDictionary<TId, T>();
        private ConcurrentDictionary<TId, ISaveAction> SaveActions = new ConcurrentDictionary<TId, ISaveAction>();

        public SnoCollection(SnoDatabase database, string name, Func<T, TId> idSelector)
        {
            Database = database;
            Name = name;
            IdSelector = idSelector;
        }

        public void Load(IEnumerable<ZipEntry> entries)
        {
            foreach (var entry in entries)
            {
                string itemJson;
                using (var reader = new StreamReader(entry.OpenReader()))
                {
                    itemJson = reader.ReadToEnd();
                }

                var item = JsonConvert.DeserializeObject<T>(itemJson, SnoDbConfig.DefaultSerializerSettings);
                var id = IdSelector(item);
                Items[id] = item;
            }
        }

        public IQueryable<T> Query()
        {
            return Items.Values.AsQueryable();
        }

        public void AddOrReplace(T item)
        {
            var id = IdSelector(item);
            Items[id] = item;
            var itemPath = GetItemPath(item);
            SaveActions[id] = new AddOrReplaceSaveAction(itemPath, item);

            if (SnoDbConfig.PersistenceMode == PersistenceMode.Instant)
                Database.Save();
        }

        public string GetItemPath(T item)
        {
            var id = IdSelector(item);
            return $"{Name}/{id}.json";
        }

        public void Remove(T item)
        {
            var id = IdSelector(item);
            Items.TryRemove(id, out item);

            var itemPath = GetItemPath(item);
            SaveActions[id] = new RemoveSaveAction(itemPath);

            if (SnoDbConfig.PersistenceMode == PersistenceMode.Instant)
                Database.Save();
        }

        public void Save(ZipFile archive)
        {
            foreach (var saveAction in SaveActions.Values)
            {
                saveAction.Save(archive);
            }

            SaveActions.Clear();
        }
    }
}
